reshuffle.go

```diff
@@ -28,12 +38,47 @@ func Reshuffle(slice Slice) Slice {
 	if err := canMakeCombiningFrame(slice); err != nil {
 		typecheck.Panic(1, err.Error())
 	}
-	return &reshuffleSlice{makeName("reshuffle"), slice}
+	return &reshuffleSlice{makeName("reshuffle"), nil, slice}
+}
+
+// Repartition (re-)partitions the slice according to the provided function
+// fn, which is invoked for each record in the slice to assign that record's
+// shard. The function is is supplied with the number of shards to partition
+// over as well as the column values; the assigned shard is returned.
+//
+// Schematically:
+//
+//	Repartition(Slice<t1, t2, ..., tn> func(nshard int, v1 t1, ..., vn tn) int)  Slice<t1, t2, ..., tn>
+func Repartition(slice Slice, fn interface{}) Slice {
+	var (
+		expectArg = slicetype.Append(sliceTypeInt, slice)
+		expectRet = sliceTypeInt
+	)
+	arg, ret, ok := typecheck.Func(fn)
+	if !ok {
+		typecheck.Panicf(1, "repartition: not a function: %T", fn)
+	}
+	if !typecheck.Equal(arg, expectArg) || !typecheck.Equal(ret, expectRet) {
+		typecheck.Panicf(1, "repartiton: expected %s, got %T", slicetype.Signature(expectArg, expectRet), fn)
+	}
+	fval := slicefunc.Of(fn)
+	part := func(frame frame.Frame, nshard int, shards []int) {
+		args := make([]reflect.Value, slice.NumOut()+1)
+		args[0] = reflect.ValueOf(nshard)
+		for i := range shards {
+			for j := 0; j < slice.NumOut(); j++ {
+				args[j+1] = frame.Index(j, i)
+			}
+			result := fval.Call(context.TODO(), args)
```

Comment:
for the context.TODO(), is it your intention to use a context as for counters here too? That is, if a context is provided, then use it in the Call also. If so, probably a good idea to say so and to add an issue for cleaning these up in the future.

Reply:
yes, that's the plan: it requires some internal changes to plumb through contexts. i'll add a note.

Reply:
actually it was just quicker to fix it right away.